File: nanobot-channel-webui/src/nanobot_channel_webui/tenant_runtime/guard.py
"""Guard SDK for business skills and command wrappers."""

from __future__ import annotations

import json
import os
from dataclasses import dataclass
from pathlib import Path
from typing import Any

from .contracts import TenantPolicy


class TenantAccessDenied(PermissionError):
    """Raised when a scoped user attempts to access unauthorized data."""


@dataclass(frozen=True, slots=True)
class GuardDecision:
    allowed: bool
    reason: str


class TenantGuard:
    """Small SDK used by business skills to enforce injected tenant policy."""

    def __init__(self, policy: TenantPolicy) -> None:
        self.policy = policy

    @classmethod
    def from_payload(cls, payload: dict[str, Any]) -> "TenantGuard":
        return cls(TenantPolicy.from_payload(payload))

    @classmethod
    def from_file(cls, path: str | Path) -> "TenantGuard":
        payload = json.loads(Path(path).read_text(encoding="utf-8"))
        return cls.from_payload(payload)

    @classmethod
    def from_environment(cls, env: dict[str, str] | None = None) -> "TenantGuard":
        source = env or os.environ
        policy_file = source.get("NANOBOT_WEBUI_POLICY_FILE", "").strip()
        if policy_file:
            return cls.from_file(policy_file)
        payload = {
            "user_id": source.get("NANOBOT_WEBUI_USER_ID", ""),
            "email": source.get("NANOBOT_WEBUI_USER_EMAIL", ""),
            "role": source.get("NANOBOT_WEBUI_ROLE", "admin"),
            "business_role": source.get("NANOBOT_WEBUI_BUSINESS_ROLE", "admin"),
            "resources": [],
            "scopes": {},
        }
        return cls.from_payload(payload)

    @property
    def is_unrestricted(self) -> bool:
        return self.policy.is_unrestricted

    def decide(
        self,
        resource: str,
        action: str,
        *,
        scope_key: str = "",
        scope_value: str = "",
    ) -> GuardDecision:
        if self.policy.allows(resource, action, scope_key=scope_key, scope_value=scope_value):
            return GuardDecision(True, "allowed")
        return GuardDecision(False, "tenant_scope_denied")

    def require(
        self,
        resource: str,
        action: str,
        *,
        scope_key: str = "",
        scope_value: str = "",
    ) -> None:
        decision = self.decide(resource, action, scope_key=scope_key, scope_value=scope_value)
        if not decision.allowed:
            raise TenantAccessDenied(decision.reason)

    def filter_scope_values(self, resource: str, action: str, values: list[str], *, scope_key: str) -> list[str]:
        return [
            value
            for value in values
            if self.policy.allows(resource, action, scope_key=scope_key, scope_value=value)
        ]

    def scope_rows(
        self,
        resource: str,
        action: str,
        rows: list[dict[str, Any]],
        *,
        scope_key: str,
        field: str | None = None,
    ) -> list[dict[str, Any]]:
        """Filter dictionaries by one scoped business field.

        Business skills should use this after loading data and before returning
        results. The field defaults to `scope_key`, so a company scope expects a
        `company` field unless the skill passes a domain-specific field name.
        """

        source_field = field or scope_key
        return [
            row
            for row in rows
            if self.policy.allows(
                resource,
                action,
                scope_key=scope_key,
                scope_value=str(row.get(source_field) or ""),
            )
        ]

    def allowed_scope_values(self, resource: str, scope_key: str) -> tuple[str, ...]:
        return self.policy.scope_values(resource, scope_key)

File: nanobot-channel-webui/src/nanobot_channel_webui/tenant_runtime/guard.mjs
import fs from "node:fs";

export class TenantAccessDenied extends Error {
  constructor(reason = "tenant_scope_denied") {
    super(reason);
    this.name = "TenantAccessDenied";
    this.code = reason;
  }
}

function asList(value) {
  if (value == null) return [];
  if (Array.isArray(value)) return value.map((item) => String(item).trim()).filter(Boolean);
  if (typeof value === "string") {
    return value
      .replaceAll("，", ",")
      .split(",")
      .map((item) => item.trim())
      .filter(Boolean);
  }
  return [String(value).trim()].filter(Boolean);
}

export function loadPolicyFromFile(path) {
  return JSON.parse(fs.readFileSync(path, "utf8"));
}

export function loadPolicyFromEnv(env = process.env) {
  const policyFile = String(env.NANOBOT_WEBUI_POLICY_FILE || "").trim();
  if (policyFile) return loadPolicyFromFile(policyFile);
  return {
    version: "tenant-runtime/v1",
    subject: {
      user_id: String(env.NANOBOT_WEBUI_USER_ID || ""),
      email: String(env.NANOBOT_WEBUI_USER_EMAIL || ""),
      role: String(env.NANOBOT_WEBUI_ROLE || "user"),
      business_role: String(env.NANOBOT_WEBUI_BUSINESS_ROLE || "scoped"),
    },
    resources: [],
    scopes: {},
  };
}

export function isUnrestricted(policy) {
  const subject = policy?.subject || policy || {};
  return subject.role === "admin" || subject.business_role === "admin" || subject.businessRole === "admin";
}

function resourcePermission(policy, resource) {
  if (isUnrestricted(policy)) return { resource, actions: ["*"], scopes: [{ key: "*", values: ["*"] }] };
  return (Array.isArray(policy?.resources) ? policy.resources : []).find((item) => item?.resource === resource) || null;
}

export function scopeValues(policy, resource, scopeKey) {
  if (isUnrestricted(policy)) return ["*"];
  const permission = resourcePermission(policy, resource);
  const scope = (permission?.scopes || []).find((item) => item?.key === scopeKey);
  return asList(scope?.values);
}

export function allows(policy, { resource, action, scopeKey = "", scopeValue = "" }) {
  if (isUnrestricted(policy)) return true;
  const permission = resourcePermission(policy, resource);
  if (!permission) return false;
  const actions = asList(permission.actions);
  if (!actions.includes("*") && !actions.includes(action)) return false;
  if (!scopeKey) return true;
  const values = scopeValues(policy, resource, scopeKey);
  if (values.includes("*")) return true;
  const normalizedValue = String(scopeValue || "").trim();
  return Boolean(normalizedValue) && values.includes(normalizedValue);
}

export function assertAllowed(policy, requirement) {
  if (!allows(policy, requirement)) throw new TenantAccessDenied("tenant_scope_denied");
}

export function filterAllowedValues(policy, resource, action, values, { scopeKey }) {
  return values.filter((value) => allows(policy, { resource, action, scopeKey, scopeValue: value }));
}

export function scopeRows(policy, resource, action, rows, { scopeKey, field = scopeKey, getValue } = {}) {
  return rows.filter((row) => {
    const value = getValue ? getValue(row) : row?.[field];
    return allows(policy, { resource, action, scopeKey, scopeValue: value });
  });
}
